pkg/offsetguess/offsetguess.go

```diff
@@ -0,0 +1,335 @@
+// where, for each field to guess, we store the value of
+//     (struct sock *)skp + possible_offset
+// in the eBPF map. Then, back in userspace (checkAndUpdateCurrentOffset()), we
+// check that value against the expected value of the field, advancing the
+// offset and repeating the process until we find the value we expect. Then, we
+// guess the next field.
+func Guess(b *elf.Module) error {
+	listenIP := "127.0.0.2"
+	listenPort := uint16(9091)
+	bindAddress := fmt.Sprintf("%s:%d", listenIP, listenPort)
+
+	currentNetns, err := ownNetNS()
+	if err != nil {
+		return fmt.Errorf("error getting current netns: %v", err)
+	}
+
+	mp := b.Map("tcptracer_status")
+
+	var zero uint64
+	pidTgid := uint64(os.Getpid()<<32 | syscall.Gettid())
+
+	status := &tcpTracerStatus{
+		status:  checking,
+		pidTgid: pidTgid,
+	}
+
+	// if we already have the offsets, just return
+	err = b.LookupElement(mp, unsafe.Pointer(&zero), unsafe.Pointer(status))
+	if err == nil && status.status == ready {
+		return nil
+	}
+
+	listenCompleted := make(chan struct{})
+	closeListen := make(chan struct{})
+
+	go listen(bindAddress, "tcp4", listenCompleted, closeListen)
+	<-listenCompleted
+	defer close(closeListen)
+
+	if err := b.UpdateElement(mp, unsafe.Pointer(&zero), unsafe.Pointer(status), 0); err != nil {
+		return fmt.Errorf("error initializing tcptracer_status map: %v", err)
+	}
+
+	expected := &fieldValues{
+		// 127.0.0.1
+		saddr: 0x0100007F,
+		// 127.0.0.2
+		daddr: 0x0200007F,
+		// will be set later
+		sport:  0,
+		dport:  htons(listenPort),
+		netns:  uint32(currentNetns),
+		family: syscall.AF_INET,
+	}
+
+	for status.status != ready {
+		// for ipv6, we don't need the source port because we already guessed
+		// it doing ipv4 connections so we use a random destination address and
+		// try to connect to it
+		expected.daddrIPv6[0] = rand.Uint32()
+		expected.daddrIPv6[1] = rand.Uint32()
+		expected.daddrIPv6[2] = rand.Uint32()
+		expected.daddrIPv6[3] = rand.Uint32()
+
+		ip := ipv6FromUint32Arr(expected.daddrIPv6)
+
+		if status.what != guessDaddrIPv6 {
+			conn, err := net.Dial("tcp4", bindAddress)
+			if err != nil {
+				return fmt.Errorf("error dialing %q: %v\n", bindAddress, err)
+			}
+
+			// get the source port assigned by the kernel
+			sport, err := strconv.Atoi(strings.Split(conn.LocalAddr().String(), ":")[1])
+			if err != nil {
+				return fmt.Errorf("error converting source port: %v", err)
+			}
+
+			expected.sport = htons(uint16(sport))
+
+			// set SO_LINGER to 0 so the connection state after closing is
+			// CLOSE instead of TIME_WAIT. In this way, they will disappear
+			// from the conntrack table after around 10 seconds instead of 2
+			// minutes
+			if tcpConn, ok := conn.(*net.TCPConn); ok {
+				tcpConn.SetLinger(0)
+			} else {
+				return fmt.Errorf("not a tcp connection unexpectedly")
+			}
+
+			conn.Close()
+		} else {
+			conn, err := net.Dial("tcp6", fmt.Sprintf("[%s]:9092", ip))
+			if err == nil {
```

Comment:
I'm writing a comment to explain it:

```
// Since we connect to a random IP, this will most likely fail.
// In the unlikely case where it connects successfully, we close
// the connection to avoid a leak.
```